src/plugins/profiling/common/topn.ts

```diff
@@ -13,42 +13,74 @@ import {
 
 import { StackFrameMetadata } from './profiling';
 
-type TopNBucket = {
-  Value: string;
+export type TopNSample = {
+  Timestamp: number;
   Count: number;
+  Category: string;
 };
 
-type TopNBucketsByDate = {
-  TopN: Record<number, TopNBucket[]>;
+export type TopNSamples = {
+  TopN: TopNSample[];
 };
 
-type TopNContainers = TopNBucketsByDate;
-type TopNDeployments = TopNBucketsByDate;
-type TopNHosts = TopNBucketsByDate;
-type TopNThreads = TopNBucketsByDate;
+type TopNContainers = TopNSamples;
+type TopNDeployments = TopNSamples;
+type TopNHosts = TopNSamples;
+type TopNThreads = TopNSamples;
 
-type TopNTraces = TopNBucketsByDate & {
+type TopNTraces = TopNSamples & {
   Metadata: Record<string, StackFrameMetadata[]>;
 };
 
 type TopN = TopNContainers | TopNDeployments | TopNHosts | TopNThreads | TopNTraces;
 
-export function createTopNBucketsByDate(
-  histogram: AggregationsHistogramAggregate
-): TopNBucketsByDate {
-  const topNBucketsByDate: Record<number, TopNBucket[]> = {};
+export function createTopNSamples(histogram: AggregationsHistogramAggregate): TopNSample[] {
+  const buckets = new Map();
+  const uniqueCategories = new Set<string>();
 
   const histogramBuckets = (histogram?.buckets as AggregationsHistogramBucket[]) ?? [];
   for (let i = 0; i < histogramBuckets.length; i++) {
-    const key = histogramBuckets[i].key / 1000;
-    topNBucketsByDate[key] = [];
+    const counts = new Map();
     histogramBuckets[i].group_by.buckets.forEach((item: any) => {
-      topNBucketsByDate[key].push({
-        Value: item.key,
-        Count: item.count.value,
-      });
+      uniqueCategories.add(item.key);
+      counts.set(item.key, item.count.value);
     });
+    buckets.set(histogramBuckets[i].key, counts);
   }
```

Comment:
Unfortunately a few constraints made it difficult to simplify the existing code:

1. `histogram?.buckets` can be undefined so I needed to keep the nullish coalescing operator to avoid using `try/catch`
2. We need to extract three values from `histogram?.buckets` (timestamp, count, category). To do this, we have to follow `histogram?.buckets` down two levels to get all required values. The code using `for` loops versus a functional approach was roughly the same length and complexity.